### PR TITLE
Update configuration of CircleCI builds

### PR DESCRIPTION
**Pull request by QuLogic**
## PR Summary

I noticed that CircleCI didn't seem to be doing any caching. This was because `save_cache` and `restore_cache` were _on_ the `run` step, but really should have been separate steps.

Before doing that however, I re-wrote the config to use CircleCI's new re-usable command style (as opposed to using YAML anchors). Since this allows multiple steps in each named command, it allows removing some duplicated steps and putting the cache in the right place. Finally, I added a cache for Sphinx's doctrees. This is not intended to provide much speedup (it's about 30s out of 7m for me), but rather to hold on to the intersphinx mappings because we seem to fail to download them periodically.

These are in separate commits for _slightly_ easier review.

## PR Checklist

- [N/A] Has Pytest style unit tests
- [N/A] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [N/A] New features are documented, with examples if plot related
- [N/A] Documentation is sphinx and numpydoc compliant
- [N/A] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [N/A] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by QuLogic** [comment]
Hmm, I may have to switch to the re-usable config file format instead of YAML anchors.

**Comment by QuLogic** [comment]
Aha, success, now I actually see a cache save and restore. I can't decide if we should re-write all anchors into the re-usable config format like the fonts step, just change the one, or hoist cache stuff into the individual `jobs`. Note, I intend to add another cache to the docs step, which, for the second option, would move things around to fit into the `commands` key.

**Comment by QuLogic** [comment]
I opted to just convert everything; see separate commits for individual changes, and the initial post for a new description of what this does.

File: .circleci/config.yml [hunk 1]
```diff
@@ -133,18 +156,16 @@ jobs:
...
-      - run: *mpl-install
+      - deps-install
+      - mpl-install
 
-      - run: *doc-build
+      - doc-build
 
-      - run: *doc-bundle
-      - store_artifacts:
-          path: doc/build/sphinx-gallery-files.tar.gz
```
Comment (tacaswell) [inline .circleci/config.yml:147]: Do we need this artifact?

Comment (QuLogic) [reply .circleci/config.yml:147]: I'm not sure; it was originally there to confirm that sphinx-gallery was working, but that seems rather rare these days. On the other hand, it really doesn't spend much time doing this at all.

**Pull request merged**